src/clients/Deluge.ts

```diff
@@ -0,0 +1,232 @@
+			);
+		}
+		return json;
+	}
+
+	/**
+	 * checks enabled plugins for "Label"
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.call({
+			method: "core.get_enabled_plugins",
+			params: [],
+		});
+		return enabledLabels.result && enabledLabels.result.includes("Label");
+	}
+
+	/**
+	 * if Label plugin is loaded, adds (if necessary)
+	 * and sets the label based on torrent hash.
+	 */
+	private async setLabel(infoHash: string): Promise<void> {
+		if (this.isLabelEnabled) {
+			const setResult = await this.call({
+				method: "label.set_torrent",
+				params: [infoHash, this.delugeLabel],
+			});
+			if (
+				setResult.error &&
+				setResult.error.code &&
+				setResult.error.code == 4
+			) {
+				await this.call({
+					method: "label.add",
+					params: [this.delugeLabel],
+				});
+				await this.call({
+					method: "label.set_torrent",
+					params: [infoHash, this.delugeLabel],
+				});
+			}
+		}
+	}
+
+	/**
+	 * injects a torrent into deluge client
+	 */
+	async inject(
+		newTorrent: Metafile,
+		searchee: Searchee,
+		path?: string
+	): Promise<InjectionResult> {
+		if (searchee.infoHash && !(await this.checkCompleted(searchee))) {
+			return InjectionResult.TORRENT_NOT_COMPLETE;
+		}
+		const params = this.formatData(
+			`${newTorrent.name}.cross-seed.torrent`,
+			newTorrent.encode().toString("base64"),
+			path
+		);
+		const addResult = await this.call({
+			method: "core.add_torrent_file",
+			params: params,
+		});
+		if (addResult.result) {
+			this.setLabel(newTorrent.infoHash);
+			return InjectionResult.SUCCESS;
+		} else if (
+			addResult.error &&
+			addResult.error.message &&
+			addResult.error.message.includes("already")
+		) {
+			return InjectionResult.ALREADY_EXISTS;
+		} else if (addResult.error && addResult.error.message) {
+			logger.debug({
+				label: Label.DELUGE,
+				message: `injection failed: ${addResult.error.message}`,
+			});
+			return InjectionResult.FAILURE;
+		}
+	}
+
+	/**
+	 * formats the json for rpc calls to inject
+	 */
+	private formatData(filename: string, filedump: string, path: string) {
+		return [
+			filename,
+			filedump,
+			{
+				add_paused: false,
+				seed_mode: getRuntimeConfig().skipRecheck,
+				download_location: path,
+			},
+		];
+	}
+
+	/**
+	 * returns true if the torrent hash is Seeding (completed)
+	 */
+	async checkCompleted(searchee: Searchee): Promise<boolean> {
+		try {
+			const params = [["state", "progress"], { hash: searchee.infoHash }];
+
+			const response = await this.call({
+				method: "web.update_ui",
+				params: params,
+			});
+			return (
+				response.result &&
+				response.result.torrents &&
+				(response.result.torrents[searchee.infoHash].state ===
+					"Seeding" ||
+					(response.result.torrents[searchee.infoHash].state ===
+						"Paused" &&
+						response.result.torrents[searchee.infoHash].progress ===
+							100))
```

Comment:
do we really need to check `state` if `progress` is 100? maybe we only need to check progress?

Reply:
progress only returns if the state is not "Seeding" - so calling both will only return when progress is incomplete (while downloading or paused) - If paused, we cannot assume 100%, but it is an entirely possible scenario if seeding goals are set and met for that specific torrent on its tracker, therefore we would need to check progress agnostically, in that case, to cross-seed only completed.